### PR TITLE
fix data races in concurrent use of MockTransport

### PR DESCRIPTION
**Pull request by Pomyk**
Allows concurrent use of http client with MockTransport.
Parallel tests using Activate/Deactivate are not safe.

### CODE REVIEW

File: transport.go [hunk 1]
```diff
@@ -21,13 +22,20 @@ func ConnectionFailure(*http.Request) (*http.Response, error) {
 
 // NewMockTransport creates a new *MockTransport with no responders.
 func NewMockTransport() *MockTransport {
-	return &MockTransport{make(map[string]Responder), nil, make(map[string]int), 0}
+	return &MockTransport{
+		mu:             &sync.RWMutex{},
```
Comment (SchumacherFM) [inline transport.go]: No need for a pointer to RWMutex. Remove the line entirely.

File: transport.go [hunk 2]
```diff
@@ -21,13 +22,20 @@ func ConnectionFailure(*http.Request) (*http.Response, error) {
...
+		responders:     make(map[string]Responder),
+		noResponder:    nil,
```
Comment (SchumacherFM) [inline transport.go]: Line not needed. Can be removed.

File: transport.go [hunk 3]
```diff
@@ -21,13 +22,20 @@ func ConnectionFailure(*http.Request) (*http.Response, error) {
...
+		callCountInfo:  make(map[string]int),
+		totalCallCount: 0,
```
Comment (SchumacherFM) [inline transport.go]: Line not needed. Can be removed.

File: transport.go [hunk 4]
```diff
@@ -21,13 +22,20 @@ func ConnectionFailure(*http.Request) (*http.Response, error) {
...
+	}
 }
 
 // MockTransport implements http.RoundTripper, which fulfills single http requests issued by
 // an http.Client.  This implementation doesn't actually make the call, instead deferring to
 // the registered list of responders.
 type MockTransport struct {
+	mu             *sync.RWMutex
```
Comment (SchumacherFM) [inline transport.go]: Remove `*` because no pointer needed

File: transport.go [hunk 5]
```diff
@@ -137,29 +148,41 @@ func (m *MockTransport) responderForKey(key string) Responder {
...
+	m.mu.Lock()
 	m.noResponder = responder
+	m.mu.Unlock()
 }
 
 // Reset removes all registered responders (including the no responder) from the MockTransport
 func (m *MockTransport) Reset() {
+	m.mu.Lock()
 	m.responders = make(map[string]Responder)
 	m.noResponder = nil
```
Comment (SchumacherFM) [inline transport.go]: Not needed, default value is already nil

**Comment by SchumacherFM** [comment]
Sorry for commenting ;-) Couldn't resists. Otherwise looks fine.

File: transport.go (hunk 5 above)
Comment (Pomyk) [inline transport.go]: I think this one should stay :)

**Comment by Pomyk** [comment]
Thanks for review, code updated.

**Comment by DocMerlin** [comment]
This is great. :-)

**Comment by dlebech** [comment]
Looks good to me. @SchumacherFM do you have any further comments before we merge?

**Comment by DocMerlin** [comment]
This is funny, I literally just ran into this bug 17 minutes ago.
Thanks for being on top of things.

**Comment by SchumacherFM** [comment]
Perfect. Lets merge!

**Comment by dlebech** [comment]
Thanks for the PR @Pomyk ! :1st_place_medal:

**Pull request merged**